File: implementations/java/provekit-java-contract-annotations/src/main/java/com/provekit/contract/RealizationTags.java
package com.provekit.contract;

import com.provekit.ir.Jcs;
import java.util.Objects;

/** Builder-style tagging primitives for per-concept realization metadata. */
public final class RealizationTags {
    private RealizationTags() {}

    /** A tagged realization memento matching the substrate enum shape. */
    public sealed interface RealizationMemento
            permits FirstClass, Composition, Boundary, SugarCarrier {
        Jcs.Json toJson();

        default String toJcsString() {
            return Jcs.encode(toJson());
        }

        default String recomputeCid() {
            return Jcs.cid(toJson());
        }
    }

    /** A first-class language-native concept realization. */
    public record FirstClass(String syntacticPattern, String surfaceLocator)
            implements RealizationMemento {
        public FirstClass {
            syntacticPattern = requireText(syntacticPattern, "syntacticPattern");
            surfaceLocator = requireText(surfaceLocator, "surfaceLocator");
        }

        @Override
        public Jcs.Json toJson() {
            return Jcs.object(
                    "kind", Jcs.string("first-class"),
                    "surface_locator", Jcs.string(surfaceLocator),
                    "syntactic_pattern", Jcs.string(syntacticPattern));
        }
    }

    /** A realization expressed as a content-addressed concept composition tree. */
    public record Composition(String compositionTreeCid) implements RealizationMemento {
        public Composition {
            compositionTreeCid = requireText(compositionTreeCid, "compositionTreeCid");
        }

        @Override
        public Jcs.Json toJson() {
            return Jcs.object(
                    "composition_tree_cid", Jcs.string(compositionTreeCid),
                    "kind", Jcs.string("composition"));
        }
    }

    /** A library or API boundary realization. */
    public record Boundary(String library, String api, String boundaryContractCid)
            implements RealizationMemento {
        public Boundary {
            library = requireText(library, "library");
            api = requireText(api, "api");
            boundaryContractCid = requireText(boundaryContractCid, "boundaryContractCid");
        }

        @Override
        public Jcs.Json toJson() {
            return Jcs.object(
                    "api", Jcs.string(api),
                    "boundary_contract_cid", Jcs.string(boundaryContractCid),
                    "kind", Jcs.string("boundary"),
                    "library", Jcs.string(library));
        }
    }

    /** A realization carried implicitly by concept-citation comment sugar. */
    public record SugarCarrier() implements RealizationMemento {
        @Override
        public Jcs.Json toJson() {
            return Jcs.object("kind", Jcs.string("sugar-carrier"));
        }
    }

    /**
     * Tag a concept op with a language-native syntactic form.
     *
     * <pre>{@code
     * RealizationTags.RealizationMemento realization =
     *         RealizationTags.tagFirstClass("concept:add", "${x} + ${y}", "binary-operator");
     * // Returns RealizationTags.FirstClass
     * }</pre>
     */
    public static RealizationMemento tagFirstClass(
            String conceptOp, String syntacticPattern, String surfaceLocator) {
        requireText(conceptOp, "conceptOp");
        return new FirstClass(syntacticPattern, surfaceLocator);
    }

    /**
     * Tag a concept op with a content-addressed composition tree.
     *
     * <pre>{@code
     * RealizationTags.RealizationMemento realization =
     *         RealizationTags.tagComposition("concept:list", compositionTreeCid);
     * // Returns RealizationTags.Composition
     * }</pre>
     */
    public static RealizationMemento tagComposition(String conceptOp, String compositionTree) {
        requireText(conceptOp, "conceptOp");
        return new Composition(compositionTree);
    }

    /**
     * Tag a concept op with a library or API boundary contract.
     *
     * <pre>{@code
     * RealizationTags.RealizationMemento realization =
     *         RealizationTags.tagBoundary(
     *                 "concept:http-request",
     *                 "python-requests",
     *                 "requests.get",
     *                 boundaryContractCid);
     * // Returns RealizationTags.Boundary
     * }</pre>
     */
    public static RealizationMemento tagBoundary(
            String conceptOp, String library, String api, String boundaryContractCid) {
        requireText(conceptOp, "conceptOp");
        return new Boundary(library, api, boundaryContractCid);
    }

    /**
     * Tag a concept op as a concept-citation sugar carrier.
     *
     * <pre>{@code
     * RealizationTags.RealizationMemento realization =
     *         RealizationTags.tagSugarCarrier("concept:free");
     * // Returns RealizationTags.SugarCarrier
     * }</pre>
     */
    public static RealizationMemento tagSugarCarrier(String conceptOp) {
        requireText(conceptOp, "conceptOp");
        return new SugarCarrier();
    }

    private static String requireText(String value, String field) {
        String text = Objects.requireNonNull(value, field);
        if (text.isEmpty()) {
            throw new IllegalArgumentException(field + " must be non-empty");
        }
        return text;
    }
}

File: implementations/rust/provekit-ir-types/src/realization_tags.rs
// SPDX-License-Identifier: Apache-2.0

use crate::{
    BoundaryRealization, CompositionRealization, FirstClassRealization, RealizationMemento,
    SugarCarrierRealization,
};

/// Tag a concept op with a language-native syntactic form.
///
/// # Example
///
/// ```
/// use provekit_ir_types::realization_tags::tag_first_class;
///
/// let realization = tag_first_class(
///     "concept:add",
///     "${x} + ${y}",
///     "binary-operator",
/// );
/// // Returns RealizationMemento::FirstClass(...)
/// ```
pub fn tag_first_class(
    concept_op: &str,
    syntactic_pattern: &str,
    surface_locator: &str,
) -> RealizationMemento {
    let _ = concept_op;
    RealizationMemento::FirstClass(FirstClassRealization {
        syntactic_pattern: syntactic_pattern.to_string(),
        surface_locator: surface_locator.to_string(),
    })
}

/// Tag a concept op with a content-addressed composition tree.
///
/// # Example
///
/// ```
/// use provekit_ir_types::realization_tags::tag_composition;
///
/// let realization = tag_composition(
///     "concept:list",
///     "blake3-512:11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111",
/// );
/// // Returns RealizationMemento::Composition(...)
/// ```
pub fn tag_composition(concept_op: &str, composition_tree: &str) -> RealizationMemento {
    let _ = concept_op;
    RealizationMemento::Composition(CompositionRealization {
        composition_tree_cid: composition_tree.to_string(),
    })
}

/// Tag a concept op with a library or API boundary contract.
///
/// # Example
///
/// ```
/// use provekit_ir_types::realization_tags::tag_boundary;
///
/// let realization = tag_boundary(
///     "concept:http-request",
///     "python-requests",
///     "requests.get",
///     "blake3-512:22222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222",
/// );
/// // Returns RealizationMemento::Boundary(...)
/// ```
pub fn tag_boundary(
    concept_op: &str,
    library: &str,
    api: &str,
    boundary_contract_cid: &str,
) -> RealizationMemento {
    let _ = concept_op;
    RealizationMemento::Boundary(BoundaryRealization {
        library: library.to_string(),
        api: api.to_string(),
        boundary_contract_cid: boundary_contract_cid.to_string(),
    })
}

/// Tag a concept op as a concept-citation sugar carrier.
///
/// # Example
///
/// ```
/// use provekit_ir_types::realization_tags::tag_sugar_carrier;
///
/// let realization = tag_sugar_carrier("concept:free");
/// // Returns RealizationMemento::SugarCarrier(...)
/// ```
pub fn tag_sugar_carrier(concept_op: &str) -> RealizationMemento {
    let _ = concept_op;
    RealizationMemento::SugarCarrier(SugarCarrierRealization {})
}
